pkg/clusteragent/admission/mutate/autoinstrumentation/mutators.go

```diff
@@ -0,0 +1,168 @@
+	mutatePod(*corev1.Pod) error
+}
+
+// podMutatorFunc is a podMutator as a function.
+type podMutatorFunc func(*corev1.Pod) error
+
+// mutatePod implements podMutator.
+func (f podMutatorFunc) mutatePod(pod *corev1.Pod) error {
+	return f(pod)
+}
+
+// initContainer is a podMutator which adds the container to a pod as an
+// init container. It will only add the container one time based on the
+// container name.
+//
+// This has the option to both append and prepend the container to the list.
+type initContainer struct {
+	corev1.Container
+	Prepend  bool
+	Mutators []containerMutator
+}
+
+var _ podMutator = (*initContainer)(nil)
+
+// mutatePod implements podMutator for initContainer.
+func (i initContainer) mutatePod(pod *corev1.Pod) error {
+	container := i.Container
+	for _, m := range i.Mutators {
+		if err := m.mutateContainer(&container); err != nil {
+			return err
+		}
+	}
+
+	for idx, c := range pod.Spec.InitContainers {
+		if c.Name == container.Name {
+			pod.Spec.InitContainers[idx] = container
+			return nil
+		}
+	}
+
+	pod.Spec.InitContainers = appendOrPrepend(container, pod.Spec.InitContainers, i.Prepend)
+	return nil
+}
+
+// volume is a podMutator which adds the volume to a pod.
+//
+// It will only add the volume one time based on the volume name.
+type volume struct {
+	corev1.Volume
+	Prepend bool
+}
+
+var _ podMutator = (*volume)(nil)
+
+func (v volume) mount(mount corev1.VolumeMount, prepend bool) volumeMount {
+	mount.Name = v.Name
+	return volumeMount{
+		VolumeMount: mount,
+		Prepend:     prepend,
+	}
+}
+
+// mutatePod implements podMutator for volume.
+func (v volume) mutatePod(pod *corev1.Pod) error {
+	vol := v.Volume
+	for idx, i := range pod.Spec.Volumes {
+		if i.Name == v.Volume.Name {
+			pod.Spec.Volumes[idx] = vol
+			return nil
+		}
+	}
+
+	pod.Spec.Volumes = appendOrPrepend(vol, pod.Spec.Volumes, v.Prepend)
+	return nil
+}
+
+// volumeMount is a containerMutator which adds a volume mount to a container.
+//
+// It will only add the volumeMount one time based on Name and MountPath.
+type volumeMount struct {
+	corev1.VolumeMount
+	Prepend bool
+}
+
+var _ containerMutator = (*volumeMount)(nil)
+
+// mutateContainer implements containerMutator for volumeMount.
+func (v volumeMount) mutateContainer(c *corev1.Container) error {
+	mnt := v.VolumeMount
+	for idx, vol := range c.VolumeMounts {
+		if vol.Name == mnt.Name && vol.MountPath == mnt.MountPath {
+			c.VolumeMounts[idx] = mnt
+			return nil
+		}
+	}
+
+	c.VolumeMounts = appendOrPrepend(mnt, c.VolumeMounts, v.Prepend)
+	return nil
+}
+
+func appendOrPrepend[T any](item T, toList []T, prepend bool) []T {
+	if prepend {
+		return append([]T{item}, toList...)
+	}
+
+	return append(toList, item)
+}
+
+type configKeyEnvVarMutator struct {
+	configKey string
+	envKey    string
+	getVal    func(string) string
+}
+
+func (c configKeyEnvVarMutator) mutatePod(pod *corev1.Pod) error {
+	if config.Datadog().IsSet(c.configKey) {
+		if c.getVal == nil {
```

Comment:
if you're going to restructure the code to take in a function to do the assignment, you should be consistent and not special-case string assignment. getVal should never be nil.